File: python-pytest/tests/varar/hello_varar.steps.py
from varar import steps

param, stimulus, sensor = steps(lambda: {"greeting": "", "result": 0})


@stimulus("I greet {string}")
def _(state, name):
    return {**state, "greeting": f"Hello, {name}!"}


@sensor("the greeting should be {string}")
def _(state, expected):
    return state["greeting"]


@stimulus("expression `{int}+{int}`")
def _(state, a, b):
    return {**state, "result": a + b}


@sensor("evaluate to `{int}`")
def _(state, expected):
    return state["result"]

File: java-junit-gradle/src/test/java/varar/HelloVararSteps.java
package varar;

import dev.varar.State;
import dev.varar.StepDefinitions;
import dev.varar.Steps;

public final class HelloVararSteps implements StepDefinitions<HelloVararSteps.Ctx> {

    record Ctx(String greeting, int result) implements State {}

    @Override
    public void register(Steps<Ctx> s) {
        s.state(() -> new Ctx("", 0));

        s.stimulus("I greet {string}", (Ctx ctx, String name) -> new Ctx("Hello, " + name + "!", ctx.result()));
        s.sensor("the greeting should be {string}", (Ctx ctx, String expected) -> ctx.greeting());
        s.stimulus("expression `{int}+{int}`", (Ctx ctx, Integer a, Integer b) -> new Ctx(ctx.greeting(), a + b));
        s.sensor("evaluate to `{int}`", (Ctx ctx, Integer expected) -> ctx.result());
    }
}
